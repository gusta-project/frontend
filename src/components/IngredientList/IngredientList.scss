.ingredient-list {
  td {
    padding: 4px 8px;
    vertical-align: middle;
  }
}
