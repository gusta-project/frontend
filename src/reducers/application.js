import { buildActions } from 'utils';

export const types = buildActions('application', [
  'INIT_APP',
  'LOGIN_USER',
  'REQUEST_TOKEN',
  'RECEIVE_TOKEN',
  'REQUEST_CURRENT_USER',
  'RECEIVE_CURRENT_USER',
  'LOGIN_USER_SUCCESS',
  'LOGIN_USER_FAILURE',
  'LOGOUT_USER',
  'LOGOUT_USER_SUCCESS',
  'LOGOUT_USER_FAILURE',
  'REGISTER_USER',
  'REGISTER_USER_SUCCESS',
  'REGISTER_USER_FAILURE'
]);

const initApp = () => ({
  type: types.INIT_APP
});

const loginUser = (emailAddress, password) => ({
  type: types.LOGIN_USER,
  emailAddress,
  password
});

const requestToken = (emailAddress, password) => ({
  type: types.REQUEST_TOKEN,
  emailAddress,
  password
});

const receiveToken = (token, expiration) => ({
  type: types.RECEIVE_TOKEN,
  expiration,
  token
});

const requestCurrentUser = () => ({
  type: types.REQUEST_CURRENT_USER
});

const receiveCurrentUser = user => ({
  type: types.RECEIVE_CURRENT_USER,
  user
});

const loginUserSuccess = user => ({
  type: types.LOGIN_USER_SUCCESS,
  user
});

const loginUserFailure = error => ({
  type: types.LOGIN_USER_FAILURE,
  error
});

const logoutUser = () => ({
  type: types.LOGOUT_USER
});

const logoutUserSuccess = () => ({
  type: types.LOGOUT_USER_SUCCESS
});

const logoutUserFailure = error => ({
  type: types.LOGOUT_USER_FAILURE,
  error
});

const registerUser = details => ({
  type: types.REGISTER_USER,
  details
});

const registerUserSuccess = () => ({
  type: types.REGISTER_USER_SUCCESS
});

const registerUserFailure = error => ({
  type: types.REGISTER_USER_FAILURE,
  error
});

export const actions = {
  initApp,
  loginUser,
  requestToken,
  receiveToken,
  requestCurrentUser,
  receiveCurrentUser,
  loginUserSuccess,
  loginUserFailure,
  logoutUser,
  logoutUserSuccess,
  logoutUserFailure,
  registerUser,
  registerUserSuccess,
  registerUserFailure
};

export const initialState = {
  loggingIn: false,
  loggingOut: false,
  user: null,
  error: null,
  authorization: {
    accessToken: null,
    refreshToken: null,
    expiration: null
  },
  registration: {
    registering: false,
    complete: false,
    error: null
  }
};

export const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case types.LOGIN_USER:
      return {
        ...state,
        loggingIn: true
      };
    case types.LOGOUT_USER:
      return {
        ...state,
        loggingOut: true
      };
    case types.REGISTER_USER:
      return {
        ...state,
        registration: {
          ...state.registration,
          registering: true,
          details: action.details
        }
      };
    case types.RECEIVE_TOKEN:
      return {
        ...state,
        authorization: {
          accessToken: action.token,
          expiration: action.expiration
        }
      };
    case types.RECEIVE_CURRENT_USER:
      return {
        ...state,
        user: action.user
      };
    case types.LOGOUT_USER_SUCCESS:
      return {
        ...state,
        loggingOut: false,
        user: null,
        authorization: initialState.authorization
      };
    case types.LOGIN_USER_SUCCESS:
      return {
        ...state,
        loggingIn: false
      };
    case types.LOGIN_USER_FAILURE:
    case types.LOGOUT_USER_FAILURE:
      return {
        ...state,
        loggingIn: false,
        loggingOut: false,
        error: action.error
      };
    case types.REGISTER_USER_SUCCESS:
      return {
        ...state,
        registration: {
          registering: false,
          complete: true,
          error: null
        }
      };
    case types.REGISTER_USER_FAILURE:
      return {
        ...state,
        registration: {
          registering: false,
          complete: true,
          error: action.error
        }
      };
    default:
      return state;
  }
};
