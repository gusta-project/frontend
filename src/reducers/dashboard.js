import { buildActions } from 'utils';

export const types = buildActions('dashboard', [
  'REQUEST_DASHBOARD',
  'REQUEST_DASHBOARD_SUCCESS',
  'SELECT_DASHBOARD',
  'SELECT_DASHBOARD_SUCCESS',
  'REQUEST_MIGRATIONS',
  'REQUEST_MIGRATIONS_SUCCESS',
  'REQUEST_MIGRATIONS_FAILURE',
  'REQUEST_STATS',
  'REQUEST_STATS_SUCCESS',
  'REQUEST_STATS_FAILURE'
]);

const requestDashboard = () => ({
  type: types.REQUEST_DASHBOARD
});

const requestDashboardSuccess = (dashboardComponent) => ({
  type: types.REQUEST_DASHBOARD_SUCCESS,
  dashboardComponent
});

const selectDashboard = ({ name, item }) => ({
  type: types.SELECT_DASHBOARD,
  name,
  item
});

const selectDashboardSuccess = (dashboardComponent) => ({
  type: types.SELECT_DASHBOARD_SUCCESS,
  dashboardComponent
});

const requestMigrations = () => ({
  type: types.REQUEST_MIGRATIONS
});

const requestMigrationsSuccess = (migrations) => ({
  type: types.REQUEST_MIGRATIONS_SUCCESS,
  migrations
});

const requestMigrationsFailure = (error) => ({
  type: types.REQUEST_MIGRATIONS_FAILURE,
  error
});

const requestStats = () => ({
  type: types.REQUEST_STATS
});

const requestStatsSuccess = (stats) => ({
  type: types.REQUEST_STATS_SUCCESS,
  stats
});

const requestStatsFailure = (error) => ({
  type: types.REQUEST_STATS_FAILURE,
  error
});

export const actions = {
  requestDashboard,
  requestDashboardSuccess,
  selectDashboard,
  selectDashboardSuccess,
  requestMigrations,
  requestMigrationsSuccess,
  requestMigrationsFailure,
  requestStats,
  requestStatsSuccess,
  requestStatsFailure
};

export const initialState = {
  dashboardComponent: {
    name: 'Home',
    item: null
  },
  migrations: [],
  stats: {}
};

export const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case types.REQUEST_DASHBOARD_SUCCESS:
      return {
        ...state,
        dashboardComponent: action.dashboardComponent
      };
    case types.SELECT_DASHBOARD_SUCCESS:
      return {
        ...state,
        dashboardComponent: action.dashboardComponent
      };
    case types.REQUEST_MIGRATIONS_SUCCESS:
      return {
        ...state,
        migrations: action.migrations
      };
    case types.REQUEST_MIGRATIONS_FAILURE:
      return {
        ...state,
        error: action.error
      };
    case types.REQUEST_STATS_SUCCESS:
      return {
        ...state,
        stats: action.stats
      };
    case types.REQUEST_STATS_FAILURE:
      return {
        ...state,
        error: action.error
      };
    default:
      return state;
  }
};
