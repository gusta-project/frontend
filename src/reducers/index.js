import { combineReducers } from 'redux';

import {
  reducer as application,
  initialState as applicationState
} from './application';
import { reducer as profile, initialState as profileState } from './profile';
import { reducer as toast, initialState as toastState } from './toast';

export const initialState = {
  application: applicationState,
  profile: profileState,
  toast: toastState
};

export default combineReducers({
  application,
  profile,
  toast
});
