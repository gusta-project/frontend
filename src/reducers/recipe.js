import { buildActions } from 'utils';

export const types = buildActions('recipe', [
  'CREATE_RECIPE',
  'CREATE_RECIPE_SUCCESS',
  'CREATE_RECIPE_FAILURE',
  'EDIT_RECIPE',
  'EDIT_RECIPE_SUCCESS',
  'EDIT_RECIPE_FAILURE',
  'REVERT_RECIPE',
  'REVERT_RECIPE_SUCCESS',
  'REVERT_RECIPE_FAILURE',
  'REQUEST_RECIPE',
  'REQUEST_RECIPE_SUCCESS',
  'REQUEST_RECIPE_FAILURE',
  'UPDATE_RECIPE',
  'UPDATE_RECIPE_SUCCESS',
  'UPDATE_RECIPE_FAILURE',
  'SET_DESIRED_NICOTINE_STRENGTH',
  'SET_DESIRED_DILUENT_RATIO',
  'SET_DESIRED_VOLUME',
  'SET_NICOTINE_STRENGTH',
  'SET_NICOTINE_DILUENT_RATIO',
  'SET_RECIPE_NAME',
  'SET_RECIPE_INGREDIENTS',
  'SET_RECIPE_PERCENTAGES'
]);

const createRecipe = () => ({
  type: types.CREATE_RECIPE
});

const createRecipeSuccess = () => ({
  type: types.CREATE_RECIPE_SUCCESS
});

const createRecipeFailure = () => ({
  type: types.CREATE_RECIPE_FAILURE
});

const editRecipe = () => ({
  type: types.EDIT_RECIPE
});

const editRecipeSuccess = () => ({
  type: types.EDIT_RECIPE_SUCCESS
});

const editRecipeFailure = error => ({
  type: types.EDIT_RECIPE_FAILURE,
  error
});

const revertRecipe = () => ({
  type: types.REVERT_RECIPE
});

const revertRecipeSuccess = () => ({
  type: types.REVERT_RECIPE_SUCCESS
});

const revertRecipeFailure = error => ({
  type: types.REVERT_RECIPE_FAILURE,
  error
});

const requestRecipe = id => ({
  type: types.REQUEST_RECIPE,
  id
});

const requestRecipeSuccess = recipe => ({
  type: types.REQUEST_RECIPE_SUCCESS,
  recipe
});

const requestRecipeFailure = error => ({
  type: types.REQUEST_RECIPE_FAILURE,
  error
});

const updateRecipe = id => ({
  type: types.UPDATE_RECIPE,
  id
});

const updateRecipeSuccess = recipe => ({
  type: types.UPDATE_RECIPE_SUCCESS,
  recipe
});

const updateRecipeFailure = error => ({
  type: types.UPDATE_RECIPE_FAILURE,
  error
});

const setDesiredNicotineStrength = nicotineStrength => ({
  type: types.SET_DESIRED_NICOTINE_STRENGTH,
  nicotineStrength
});

const setDesiredDiluentRatio = diluentRatio => ({
  type: types.SET_DESIRED_DILUENT_RATIO,
  diluentRatio
});

const setDesiredVolume = volume => ({
  type: types.SET_DESIRED_VOLUME,
  volume
});

const setNicotineStrength = nicotineStrength => ({
  type: types.SET_NICOTINE_STRENGTH,
  nicotineStrength
});

const setNicotineDiluentRatio = diluentRatio => ({
  type: types.SET_NICOTINE_DILUENT_RATIO,
  diluentRatio
});

const setRecipeName = name => ({
  type: types.SET_RECIPE_NAME,
  name
});

const setRecipeIngredients = ingredients => ({
  type: types.SET_RECIPE_INGREDIENTS,
  ingredients
});

const setRecipePercentages = percentages => ({
  type: types.SET_RECIPE_PERCENTAGES,
  percentages
});

export const actions = {
  createRecipe,
  createRecipeSuccess,
  createRecipeFailure,
  editRecipe,
  editRecipeSuccess,
  editRecipeFailure,
  revertRecipe,
  revertRecipeSuccess,
  revertRecipeFailure,
  requestRecipe,
  requestRecipeSuccess,
  requestRecipeFailure,
  updateRecipe,
  updateRecipeSuccess,
  updateRecipeFailure,
  setDesiredNicotineStrength,
  setDesiredDiluentRatio,
  setDesiredVolume,
  setNicotineStrength,
  setNicotineDiluentRatio,
  setRecipeName,
  setRecipeIngredients,
  setRecipePercentages
};

export const initialState = {
  savePoint: null,
  editing: false,
  active: {
    id: null,
    name: '',
    ingredients: [],
    percentages: []
  },
  error: null,
  desired: {
    nicotineStrength: 6,
    diluentRatio: 0.8,
    volume: 8
  },
  settings: {
    nicotineStrength: 100,
    nicotineDiluentRatio: 1
  }
};

export const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case types.CREATE_RECIPE_SUCCESS:
      return {
        ...state,
        active: {
          id: null,
          name: '',
          ingredients: [],
          percentages: {}
        }
      };
    case types.EDIT_RECIPE_SUCCESS:
      if (state.editing) {
        return state;
      }

      return {
        ...state,
        editing: true,
        savePoint: state.active
      };
    case types.REVERT_RECIPE_SUCCESS:
      if (!state.editing) {
        return state;
      }

      return {
        ...state,
        editing: false,
        savePoint: null,
        active: state.savePoint
      };
    case types.REQUEST_RECIPE_SUCCESS:
    case types.UPDATE_RECIPE_SUCCESS:
      return {
        ...state,
        active: action.recipe
      };
    case types.EDIT_RECIPE_FAILURE:
    case types.REVERT_RECIPE_FAILURE:
    case types.REQUEST_RECIPE_FAILURE:
    case types.UPDATE_RECIPE_FAILURE:
      return {
        ...state,
        error: action.error
      };
    case types.SET_DESIRED_NICOTINE_STRENGTH:
      return {
        ...state,
        desired: {
          ...state.desired,
          nicotineStrength: action.nicotineStrength
        }
      };
    case types.SET_DESIRED_DILUENT_RATIO:
      return {
        ...state,
        desired: {
          ...state.desired,
          diluentRatio: action.diluentRatio
        }
      };
    case types.SET_DESIRED_VOLUME:
      return {
        ...state,
        desired: {
          ...state.desired,
          volume: action.volume
        }
      };
    case types.SET_NICOTINE_STRENGTH:
      return {
        ...state,
        settings: {
          ...state.settings,
          nicotineStrength: action.nicotineStrength
        }
      };
    case types.SET_NICOTINE_DILUENT_RATIO:
      return {
        ...state,
        settings: {
          ...state.settings,
          nicotineDiluentRatio: action.diluentRatio
        }
      };
    case types.SET_RECIPE_NAME:
      return {
        ...state,
        active: {
          ...state.active,
          name: action.name
        }
      };
    case types.SET_RECIPE_INGREDIENTS:
      return {
        ...state,
        active: {
          ...state.active,
          ingredients: action.ingredients
        }
      };
    case types.SET_RECIPE_PERCENTAGES:
      return {
        ...state,
        active: {
          ...state.active,
          percentages: action.percentages
        }
      };
    default:
      return state;
  }
};
