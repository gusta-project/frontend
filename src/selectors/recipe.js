import { createSelector } from 'reselect';

export const getRecipe = state => state.recipe;

export const getSavePoint = createSelector(
  getRecipe,
  recipe => recipe.savePoint
);

export const isEditing = createSelector(getRecipe, recipe => recipe.editing);

export const getActiveRecipe = createSelector(
  getRecipe,
  recipe => recipe.active
);

export const getRecipeId = createSelector(getActiveRecipe, active => active.id);

export const getRecipeName = createSelector(
  getActiveRecipe,
  active => active.name
);

export const getRecipeIngredients = createSelector(
  getActiveRecipe,
  active => active.ingredients
);

export const getRecipePercentages = createSelector(
  getActiveRecipe,
  active => active.percentages
);

export const getError = createSelector(getRecipe, recipe => recipe.error);

export const getOptions = createSelector(getRecipe, recipe => recipe.options);

export const getDesired = createSelector(getRecipe, recipe => recipe.desired);

export const getSettings = createSelector(getRecipe, recipe => recipe.settings);

export const getNicotineStrength = createSelector(
  getSettings,
  settings => settings.nicotineStrength
);

export const getNicotineDiluentRatio = createSelector(
  getSettings,
  settings => settings.nicotineDiluentRatio
);

export const getDesiredNicotineStrength = createSelector(
  getDesired,
  desired => desired.nicotineStrength
);

export const getDesiredDiluentRatio = createSelector(
  getDesired,
  desired => desired.diluentRatio
);

export const getDesiredVolume = createSelector(
  getDesired,
  desired => desired.volume
);
