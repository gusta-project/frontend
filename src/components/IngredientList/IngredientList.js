import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Button, Table, InputGroup, Form } from 'react-bootstrap';

import { actions as recipeActions } from 'reducers/recipe';

export class IngredientList extends Component {
  static propTypes = {
    actions: PropTypes.shape({
      setRecipeIngredients: PropTypes.func.isRequired,
      setRecipePercentages: PropTypes.func.isRequired
    }),
    ingredients: PropTypes.arrayOf(PropTypes.object).isRequired,
    percentages: PropTypes.arrayOf(PropTypes.number).isRequired
  };

  static defaultProps = {
    ingredients: [],
    percentages: []
  };

  constructor(props) {
    super(props);

    this.removeIngredient = this.removeIngredient.bind(this);
    this.handlePercentChange = this.handlePercentChange.bind(this);
  }

  removeIngredient(id) {
    const { ingredients, actions } = this.props;

    actions.setRecipeIngredients(
      ingredients.filter((ingredient) => ingredient.Flavor.id !== id)
    );
  }

  handlePercentChange(event) {
    const { actions, ingredients, percentages } = this.props;
    const {
      target: { name, value }
    } = event;

    const floatValue = parseFloat(value);

    if (floatValue === false) {
      return;
    }

    const index = ingredients.findIndex(
      (ingredient) => ingredient.Flavor.id === name
    );

    if (index > -1) {
      actions.setRecipePercentages({ ...percentages, [name]: floatValue });
    }
  }

  render() {
    const { ingredients } = this.props;

    if (!Array.isArray(ingredients) || ingredients.length === 0) {
      return <div>No ingredients!</div>;
    }

    return (
      <Table striped className="ingredient-list">
        <tbody>
          {ingredients.map((ingredient) => {
            const {
              flavorId: id,
              Flavor: { name, Vendor: vendor }
            } = ingredient;

            return (
              <tr key={id}>
                <td>
                  {vendor.code} {name}
                </td>
                <td className="recipe-percent">
                  <InputGroup>
                    <Form.Control
                      step="0.1"
                      name={id}
                      type="number"
                      value={ingredient.percentage}
                      onChange={this.handlePercentChange}
                      placeholder="0"
                      required
                    />
                    <InputGroup.Append>
                      <InputGroup.Text>%</InputGroup.Text>
                    </InputGroup.Append>
                  </InputGroup>
                </td>
                <td className="text-right">
                  <Button
                    size="sm"
                    className="button-animation"
                    onClick={() => this.removeIngredient(id)}
                  >
                    <FontAwesomeIcon icon="trash" size="xs" title="remove" />
                  </Button>
                </td>
              </tr>
            );
          })}
        </tbody>
      </Table>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  actions: bindActionCreators(recipeActions, dispatch)
});

export default connect(null, mapDispatchToProps)(IngredientList);
