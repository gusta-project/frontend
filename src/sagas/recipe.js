import { put, all, takeLatest, call, select } from 'redux-saga/effects';

import request from 'utils/request';
import { getUser } from 'selectors/application';
import { actions, types } from 'reducers/recipe';
import { getActiveRecipe } from 'selectors/recipe';

function* editRecipeWorker() {
  try {
    const endpoint = {
      url: '/recipe',
      method: 'POST'
    };
    const user = yield select(getUser);

    if (!user) {
      throw new Error('Not logged in!');
    }

    const recipe = yield select(getActiveRecipe);

    if (!recipe) {
      throw new Error('Not editing a recipe!');
    }

    const diluents = [];
    const flavors = [];

    const data = {
      userId: user.id,
      name: recipe.name,
      RecipesDiluents: diluents,
      RecipesFlavors: flavors
    };
    const result = yield call(request.execute, { endpoint, data });

    if (!result.success) {
      throw result.error;
    }

    yield put(actions.editRecipeSuccess());
  } catch (error) {
    yield put(actions.editRecipeFailure(error));
  }
}

function* editRecipeWatcher() {
  return yield takeLatest(types.EDIT_RECIPE, editRecipeWorker);
}

function* revertRecipeWorker() {
  yield put(actions.revertRecipeSuccess());
}

function* revertRecipeWatcher() {
  return yield takeLatest(types.REVERT_RECIPE, revertRecipeWorker);
}

function* requestRecipeWorker(id) {
  try {
    const endpoint = {
      url: `/recipe/${id}`,
      method: 'GET'
    };
    const result = yield call(request.execute, { endpoint });

    if (!result.success) {
      throw result.error;
    }

    const {
      data: { recipe }
    } = result.response;

    yield put(actions.requestRecipeSuccess(recipe));
  } catch (error) {
    yield put(actions.requestRecipeFailure(error));
  }
}

function* requestRecipeWatcher() {
  return yield takeLatest(types.REQUEST_RECIPE, requestRecipeWorker);
}

export const watchers = {
  editRecipeWatcher,
  revertRecipeWatcher,
  requestRecipeWatcher
};

export const workers = {
  editRecipeWorker,
  revertRecipeWorker,
  requestRecipeWorker
};

export default function* saga() {
  yield all(Object.values(watchers).map(watcher => watcher()));
}
