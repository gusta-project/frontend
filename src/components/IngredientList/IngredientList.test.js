import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import { useDispatch } from 'react-redux';

import { actions as recipeActions } from 'reducers/recipe';

import IngredientList from './IngredientList';

jest.mock('react-redux');

describe('<IngredientList />', () => {
  const props = {
    ingredients: [
      {
        flavorId: 'A',
        Flavor: { id: 'A', name: 'A_NAME', Vendor: { code: 'A_VENDOR' } }
      },
      {
        flavorId: 'B',
        Flavor: { id: 'B', name: 'B_NAME', Vendor: { code: 'B_VENDOR' } }
      },
      {
        flavorId: 'C',
        Flavor: { id: 'C', name: 'C_NAME', Vendor: { code: 'C_VENDOR' } }
      }
    ],
    percentages: {
      A: 0,
      B: 0,
      C: 0
    }
  };

  it('renders correctly', () => {
    const { asFragment } = render(<IngredientList {...props} />);

    useDispatch.mockReturnValue(jest.fn());

    expect(asFragment()).toMatchSnapshot();
  });

  it('calls dispatch for percentage change', () => {
    const mockDispatch = jest.fn();

    useDispatch.mockReturnValue(mockDispatch);

    const { getByTestId } = render(<IngredientList {...props} />);

    expect(getByTestId('ingredient-A-percent')).not.toBeNull();

    fireEvent.change(getByTestId('ingredient-A-percent'), {
      target: { name: 'A', value: '5.0' }
    });

    expect(mockDispatch).toHaveBeenCalledWith(
      recipeActions.setRecipePercentages({
        ...props.percentages,
        A: 5.0
      })
    );
  });

  it('calls dispatch to remove from list', () => {
    const mockDispatch = jest.fn();

    useDispatch.mockReturnValue(mockDispatch);

    const { getByTestId } = render(<IngredientList {...props} />);

    expect(getByTestId('ingredient-C-remove')).not.toBeNull();

    fireEvent.click(getByTestId('ingredient-C-remove'));

    expect(mockDispatch).toHaveBeenCalledWith(
      recipeActions.setRecipeIngredients(
        props.ingredients.filter((ingredient) => ingredient.Flavor.id !== 'C')
      )
    );
  });
});
