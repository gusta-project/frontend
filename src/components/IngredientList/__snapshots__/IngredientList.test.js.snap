// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IngredientList /> renders correctly 1`] = `
<DocumentFragment>
  <table
    class="ingredient-list table table-striped"
  >
    <tbody>
      <tr>
        <td>
          A_VENDOR A_NAME
        </td>
        <td
          class="recipe-percent"
        >
          <div
            class="input-group"
          >
            <input
              class="form-control"
              data-testid="ingredient-A-percent"
              name="A"
              placeholder="0"
              required=""
              step="0.1"
              type="number"
              value=""
            />
            <div
              class="input-group-append"
            >
              <span
                class="input-group-text"
              >
                %
              </span>
            </div>
          </div>
        </td>
        <td
          class="text-right"
        >
          <button
            class="button-animation btn btn-primary btn-sm"
            data-testid="ingredient-A-remove"
            type="button"
          >
            <i
              class="fa trash"
            />
          </button>
        </td>
      </tr>
      <tr>
        <td>
          B_VENDOR B_NAME
        </td>
        <td
          class="recipe-percent"
        >
          <div
            class="input-group"
          >
            <input
              class="form-control"
              data-testid="ingredient-B-percent"
              name="B"
              placeholder="0"
              required=""
              step="0.1"
              type="number"
              value=""
            />
            <div
              class="input-group-append"
            >
              <span
                class="input-group-text"
              >
                %
              </span>
            </div>
          </div>
        </td>
        <td
          class="text-right"
        >
          <button
            class="button-animation btn btn-primary btn-sm"
            data-testid="ingredient-B-remove"
            type="button"
          >
            <i
              class="fa trash"
            />
          </button>
        </td>
      </tr>
      <tr>
        <td>
          C_VENDOR C_NAME
        </td>
        <td
          class="recipe-percent"
        >
          <div
            class="input-group"
          >
            <input
              class="form-control"
              data-testid="ingredient-C-percent"
              name="C"
              placeholder="0"
              required=""
              step="0.1"
              type="number"
              value=""
            />
            <div
              class="input-group-append"
            >
              <span
                class="input-group-text"
              >
                %
              </span>
            </div>
          </div>
        </td>
        <td
          class="text-right"
        >
          <button
            class="button-animation btn btn-primary btn-sm"
            data-testid="ingredient-C-remove"
            type="button"
          >
            <i
              class="fa trash"
            />
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
